### PR TITLE
github: add workflows for the jobs builder

### LINKED ISSUE #343: Use the Jenkins Job Builder to manage the jobs
Currently there are many problems with the way the Jenkins jobs are managed:
 * One need to edit Jenkins internal representation of the job (in XML) in order change or create a new.
 * When a XML is updated someone needs to manually get Jenkins updated. Unless it is automated?
   * So the new job can delay to be deployed; Maybe what we have running on Jenkins is out of sync with the XML files of the repository. 
 * There are some many job variants which changing them is labour intensive and error prone.
   * Example: I want to change something on the jobs for the runtime repository (pull request, default configuration). I need to change on all the OSes otherwise they can be added inconsistencies among the variants. If eventually I need to change the same thing but in different repositories, there are more XML to edit then.

There is the [Jenkins Job Builder](https://docs.openstack.org/infra/jenkins-job-builder) (JJB) which aims to easy the management of Jenkins job. With this tool you can keep the job representation in YAML (easier to review) versioned in git, generate jobs from templates and ultimately publish them in the Jenkins instance easily. 

So what I propose is:

 * Implement the Jenkins jobs configurations using JJB
   * Use templates so many jobs can be generated from a seed configuration
 * Create a script to easy the process of updating Jenkins with new (or edited) jobs
 * Have a "auto-update" job
   * Watches the `ci` repository for changes on the jobs configurations
   * Triggers the script mentioned above, which will update the Jenkins
   (ps: need to think about possible security problems)

### PR DESCRIPTION
**Pull request by wainersm**
It is added two new github workflows in this PR:

- On pull request, check that changes didn't break the jobs builder
- On push, automatically publish to Jenkins if any job configuration changed (it requires adding secrets on the configuration of this repository)

Fixes #343

### CODE REVIEW

**Comment by wainersm** [comment]
Tested in https://github.com/wainersm/kc-ci/runs/4941967596

**Comment by wainersm** [comment]
Updates on this PR:
 - Removed a dangling "if" statement.

**Comment by wainersm** [comment]
Updates on this PR:
 - Added fe50f1fc586511b079dbccdb9ce1d7813abbae71 which implements a workflow for when a PR is merge to automatically re-publish the jobs.

**Comment by wainersm** [comment]
Set do-not-merge because I need to test the new created workflow....

**Review by jodh-intel — APPROVED** [review]
Thanks @wainersm.

lgtm

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

**Comment by wainersm** [comment]
Update to this PR:

- Basically re-wrote  https://github.com/kata-containers/ci/commit/a8f4bf5ce7e6c275a45d7d0c1add60076ff5f112

It should be all working now.

File: .github/workflows/publish_jobs.yml
Comment (jodh-intel) [inline .github/workflows/publish_jobs.yml:29]: Nit: I'd consider an `ERROR:` prefix and writing the error to `stderr`.

File: .github/workflows/publish_jobs.yml
Comment (jodh-intel) [inline .github/workflows/publish_jobs.yml:33]: Nit: You could chain these into a single call:

```suggestion
          sed -i \
              -e "s/user=XXX/user=$JENKINS_USER/" \
              -e "s/password=XXX/password=$JENKINS_PASSWORD/" \
              jjb.conf
```

**Review by jodh-intel — DISMISSED** [review]
Thanks @wainersm. A couple of nits but overall...

lgtm

File: .github/workflows/publish_jobs.yml
Comment (wainersm) [reply .github/workflows/publish_jobs.yml]: Done. I was curious if sending the output to `stderr` would make any difference (i.e. github would send the message to a seperate log file) but it seems not.

File: .github/workflows/publish_jobs.yml
Comment (wainersm) [reply .github/workflows/publish_jobs.yml]: Done. Thanks! Old habits are hard to die! :) sometimes I find myself doing `cat file | grep word` still...

**Comment by wainersm** [comment]
Updated to this PR:
 - Only addressed the latest comments from @jodh-intel

**Review by snir911 — APPROVED** [review]
LGTM, thanks!
Would it make sense to have an option to run the static check upon some comment "test-jjb".. or similar

**Review by GabyCT — APPROVED** [review]

**Pull request merged**
